// sgds-govtech-react/dist/Brand.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes);
    global.Brand = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_exports, _react, _propTypes) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _react = _interopRequireDefault(_react);
  _propTypes = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

  function Brand(props) {
    return /*#__PURE__*/_react.default.createElement("a", {
      className: "navbar-item",
      href: props.link
    }, props.img ? /*#__PURE__*/_react.default.createElement("img", {
      src: props.img,
      alt: props.name
    }) : /*#__PURE__*/_react.default.createElement("h1", null, props.name));
  }

  Brand.propTypes = {
    name: _propTypes.default.string.isRequired,
    link: _propTypes.default.string.isRequired,
    img: _propTypes.default.string
  };
  var _default = Brand;
  _exports.default = _default;
});
