// sgds-govtech-react/dist/Dropdown.js
function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types", "./Button"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"), require("./Button"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes, global.Button);
    global.Dropdown = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_exports, _react, _propTypes, _Button) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _react = _interopRequireWildcard(_react);
  _propTypes = _interopRequireDefault(_propTypes);
  _Button = _interopRequireDefault(_Button);

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

  function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

  function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

  function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

  function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

  function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

  function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

  function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

  function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

  function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

  var Dropdown = /*#__PURE__*/function (_Component) {
    _inherits(Dropdown, _Component);

    var _super = _createSuper(Dropdown);

    function Dropdown() {
      var _this;

      _classCallCheck(this, Dropdown);

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      _this = _super.call.apply(_super, [this].concat(args));

      _defineProperty(_assertThisInitialized(_this), "state", {
        showDropdown: false
      });

      _defineProperty(_assertThisInitialized(_this), "toggleDropdown", function () {
        _this.setState({
          showDropdown: !_this.state.showDropdown
        });
      });

      _defineProperty(_assertThisInitialized(_this), "setContentWrapperRef", function (node) {
        _this.contentWrapperRef = node; // Set directly on 'this' since it won't change/re-render
      });

      _defineProperty(_assertThisInitialized(_this), "hideDropdownOnClick", function (event) {
        if (_this.contentWrapperRef && !_this.contentWrapperRef.contains(event.target)) {
          _this.setState({
            showDropdown: false
          });
        }
      });

      _defineProperty(_assertThisInitialized(_this), "hideDropdownOnEscape", function (event) {
        if (event.key === "Escape") {
          _this.setState({
            showDropdown: false
          });
        }
      });

      return _this;
    }

    _createClass(Dropdown, [{
      key: "componentDidMount",
      value: function componentDidMount() {
        document.addEventListener("mousedown", this.hideDropdownOnClick);
        document.addEventListener("keydown", this.hideDropdownOnEscape);
      }
    }, {
      key: "componentWillUnmount",
      value: function componentWillUnmount() {
        document.removeEventListener("mousedown", this.hideDropdownOnClick);
        document.removeEventListener("keydown", this.hideDropdownOnEscape);
      }
    }, {
      key: "render",
      value: function render() {
        var _this2 = this;

        if (this.props.isHoverable) {
          return /*#__PURE__*/_react.default.createElement("div", {
            className: "sgds-dropdown ".concat(this.state.showDropdown ? "is-active" : ""),
            onMouseLeave: function onMouseLeave() {
              return _this2.setState({
                showDropdown: false
              });
            }
          }, /*#__PURE__*/_react.default.createElement("div", {
            className: "sgds-dropdown-trigger"
          }, /*#__PURE__*/_react.default.createElement(_Button.default, {
            type: "button",
            "aria-haspopup": "true",
            "aria-controls": "sgds-dropdown-menu",
            onMouseEnter: function onMouseEnter() {
              return _this2.setState({
                showDropdown: true
              });
            }
          }, /*#__PURE__*/_react.default.createElement("span", null, this.props.title), /*#__PURE__*/_react.default.createElement("span", {
            className: "icon"
          }, /*#__PURE__*/_react.default.createElement("span", {
            className: "sgds-icon sgds-icon-chevron-".concat(this.state.showDropdown ? "up" : "down")
          })))), /*#__PURE__*/_react.default.createElement("div", {
            className: "sgds-dropdown-menu",
            role: "menu"
          }, /*#__PURE__*/_react.default.createElement("div", {
            className: "sgds-dropdown-content"
          }, this.props.children)));
        }

        return /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-dropdown ".concat(this.state.showDropdown ? "is-active" : ""),
          ref: this.setContentWrapperRef
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-dropdown-trigger"
        }, /*#__PURE__*/_react.default.createElement(_Button.default, {
          type: "button",
          "aria-haspopup": "true",
          "aria-controls": "sgds-dropdown-menu",
          onClick: this.toggleDropdown
        }, /*#__PURE__*/_react.default.createElement("span", null, this.props.title), /*#__PURE__*/_react.default.createElement("span", {
          className: "icon"
        }, /*#__PURE__*/_react.default.createElement("span", {
          className: "sgds-icon sgds-icon-chevron-".concat(this.state.showDropdown ? "up" : "down")
        })))), /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-dropdown-menu",
          role: "menu"
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-dropdown-content",
          onClick: function onClick() {
            _this2.setState({
              showDropdown: false
            });
          }
        }, this.props.children)));
      }
    }]);

    return Dropdown;
  }(_react.Component);

  Dropdown.propTypes = {
    title: _propTypes.default.string,
    isHoverable: _propTypes.default.bool
  };
  var _default = Dropdown;
  _exports.default = _default;
});
