// sgds-govtech-react/dist/Footer.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes);
    global.Footer = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_exports, _react, _propTypes) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _react = _interopRequireDefault(_react);
  _propTypes = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

  function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

  function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

  function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

  function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

  function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

  function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

  // TODO : Improve error handling of inputs into components
  function Footer(props) {
    var getDateYear = function getDateYear(date) {
      var dateRegex = /^([0-9]|[0-2][0-9]|[3][0-1]) (Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec) (\d{4})$/i;
      var matches = dateRegex.exec(date);

      if (matches) {
        return [matches[0], matches[3]];
      } else {
        throw new Error('Invalid date format for Footer: Only dd MMM yyyy date formats accepted');
      }
    }; // TODO : Refine code in renderBottomButtons() and renderBottomRightTopSectionLinks()


    var renderBottomButtons = function renderBottomButtons() {
      var links = [];
      links.push( /*#__PURE__*/_react.default.createElement("li", {
        key: "vuln"
      }, /*#__PURE__*/_react.default.createElement("a", {
        href: "https://tech.gov.sg/report_vulnerability",
        target: "_blank",
        rel: "noopener noreferrer"
      }, "Report Vulnerability")));

      if (!props.links) {
        return;
      }

      if (props.links.privacy) {
        links.push( /*#__PURE__*/_react.default.createElement("li", {
          key: "privacy"
        }, /*#__PURE__*/_react.default.createElement("a", {
          href: props.links.privacy
        }, "Privacy")));
      }

      if (props.links.termsOfUse) {
        links.push( /*#__PURE__*/_react.default.createElement("li", {
          key: "termsOfUse"
        }, /*#__PURE__*/_react.default.createElement("a", {
          href: props.links.termsOfUse
        }, "Terms of Use")));
      }

      return links;
    };

    var renderBottomRightTopSectionLinks = function renderBottomRightTopSectionLinks() {
      var links = [];

      if (!props.links) {
        return;
      }

      if (props.links.contact) {
        links.push( /*#__PURE__*/_react.default.createElement("li", {
          key: "contact",
          className: "is-inline-block-desktop-only"
        }, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement("a", {
          href: props.links.contact,
          target: "_blank",
          rel: "noopener noreferrer"
        }, "Contact"))));
      }

      if (props.links.feedback) {
        links.push( /*#__PURE__*/_react.default.createElement("li", {
          key: "feedback",
          className: "is-inline-block-desktop-only"
        }, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement("a", {
          href: props.links.feedback,
          target: "_blank",
          rel: "noopener noreferrer"
        }, "Feedback"))));
      }

      return links;
    };

    var renderBottomFooterText = function renderBottomFooterText() {
      if (!props.date) {
        return;
      }

      var _getDateYear = getDateYear(props.date),
          _getDateYear2 = _slicedToArray(_getDateYear, 2),
          date = _getDateYear2[0],
          year = _getDateYear2[1];

      return /*#__PURE__*/_react.default.createElement("div", {
        className: "col is-12 has-text-right-desktop has-text-right-tablet has-text-left-mobile"
      }, /*#__PURE__*/_react.default.createElement("p", {
        className: "is-hidden-touch"
      }, " \xA9 ", year, " Government of Singapore. Last Updated ", date), /*#__PURE__*/_react.default.createElement("p", {
        className: "is-hidden-desktop"
      }, "\xA9 ", year, " Government of Singapore"), /*#__PURE__*/_react.default.createElement("p", {
        className: "is-hidden-desktop last-updated"
      }, "Last Updated ", date));
    };

    return /*#__PURE__*/_react.default.createElement("footer", {
      className: "sgds-footer"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "top-section"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "sgds-container ".concat(props.isFluid ? "is-fluid" : "")
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "row"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "col"
    }, /*#__PURE__*/_react.default.createElement("h5", {
      className: "has-text-white"
    }, /*#__PURE__*/_react.default.createElement("b", null, props.title)))), /*#__PURE__*/_react.default.createElement("div", {
      className: "row"
    }, props.children), /*#__PURE__*/_react.default.createElement("div", {
      className: "row"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "col is-right-desktop-only"
    }, /*#__PURE__*/_react.default.createElement("ul", null, renderBottomRightTopSectionLinks()))))), /*#__PURE__*/_react.default.createElement("div", {
      className: "bottom-section"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "sgds-container ".concat(props.isFluid ? "is-fluid" : "")
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "row is-multiline"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "col is-12"
    }, /*#__PURE__*/_react.default.createElement("ul", null, renderBottomButtons())), renderBottomFooterText()))));
  }

  ;
  Footer.propTypes = {
    title: _propTypes.default.string,
    date: _propTypes.default.string,
    links: _propTypes.default.shape({
      privacy: _propTypes.default.string,
      termsOfUse: _propTypes.default.string,
      contact: _propTypes.default.string,
      feedback: _propTypes.default.string
    }),
    isFluid: _propTypes.default.bool
  };
  var _default = Footer;
  _exports.default = _default;
});
