// sgds-govtech-react/dist/Hero.js
function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes);
    global.Hero = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_exports, _react, _propTypes) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _react = _interopRequireWildcard(_react);
  _propTypes = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

  function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

  function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

  function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

  function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

  function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

  function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

  function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

  function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

  var Hero = /*#__PURE__*/function (_Component) {
    _inherits(Hero, _Component);

    var _super = _createSuper(Hero);

    function Hero(props) {
      var _this;

      _classCallCheck(this, Hero);

      _this = _super.call(this, props);
      _this.setWrapperRef = _this.setWrapperRef.bind(_assertThisInitialized(_this));
      _this.handleClickOutside = _this.handleClickOutside.bind(_assertThisInitialized(_this));
      _this.handleSearchQuery = _this.handleSearchQuery.bind(_assertThisInitialized(_this));
      _this.state = {
        selected: null,
        show: false,
        searchQuery: ""
      };
      return _this;
    }

    _createClass(Hero, [{
      key: "componentDidMount",
      value: function componentDidMount() {
        document.addEventListener("mousedown", this.handleClickOutside);
      }
    }, {
      key: "componentWillUnmount",
      value: function componentWillUnmount() {
        document.removeEventListener("mousedown", this.handleClickOutside);
      }
    }, {
      key: "setWrapperRef",
      value: function setWrapperRef(node) {
        this.wrapperRef = node;
      }
    }, {
      key: "toggleDropdown",
      value: function toggleDropdown() {
        this.setState({
          show: !this.state.show
        });
      }
    }, {
      key: "openDropdown",
      value: function openDropdown() {
        this.setState({
          show: true
        });
      }
    }, {
      key: "closeDropdown",
      value: function closeDropdown() {
        this.setState({
          show: false
        });
      }
    }, {
      key: "handleClickOutside",
      value: function handleClickOutside(event) {
        if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
          this.setState({
            show: false
          });
        }
      }
    }, {
      key: "handleSearchQuery",
      value: function handleSearchQuery(event) {
        this.setState({
          searchQuery: event.target.value
        });
      }
    }, {
      key: "selectDropdown",
      value: function selectDropdown(name) {
        var _this2 = this;

        this.setState({
          show: false,
          selected: name
        }, function () {
          return _this2.props.selectItem(name);
        });
      }
    }, {
      key: "handleSearchButton",
      value: function handleSearchButton() {
        this.props.clickSearchButton(this.state.searchQuery);
      }
    }, {
      key: "renderDropdown",
      value: function renderDropdown() {
        var _this3 = this;

        if (this.props.showDropdown && !this.props.isHoverable) {
          return /*#__PURE__*/_react.default.createElement("div", {
            className: "sgds-dropdown is-fullwidth is-menu-centered" + ( // (this.props.isHoverable ? " is-hoverable" : "") +
            this.state.show ? " is-active" : "")
          }, /*#__PURE__*/_react.default.createElement("div", {
            className: "sgds-dropdown-trigger"
          }, /*#__PURE__*/_react.default.createElement("button", {
            className: "sgds-button is-white is-medium is-outlined",
            "aria-haspopup": "true",
            "aria-controls": "sgds-dropdown-menu",
            onClick: function onClick() {
              return _this3.openDropdown();
            }
          }, /*#__PURE__*/_react.default.createElement("span", null, /*#__PURE__*/_react.default.createElement("b", null, this.state.selected ? this.state.selected : this.props.dropdownTitle)), /*#__PURE__*/_react.default.createElement("span", {
            className: "icon"
          }, /*#__PURE__*/_react.default.createElement("span", {
            className: "sgds-icon sgds-icon-chevron-down"
          })))), this.props.dropdownItems ? /*#__PURE__*/_react.default.createElement("div", {
            className: "sgds-dropdown-menu",
            role: "menu",
            ref: this.setWrapperRef
          }, /*#__PURE__*/_react.default.createElement("div", {
            className: "sgds-dropdown-content"
          }, this.props.dropdownItems.map(function (item, index) {
            return /*#__PURE__*/_react.default.createElement("a", {
              href: "#",
              key: index,
              className: "sgds-dropdown-item",
              onClick: function onClick() {
                return _this3.selectDropdown(item);
              }
            }, item);
          }))) : "");
        } else if (this.props.showDropdown && this.props.isHoverable) {
          return /*#__PURE__*/_react.default.createElement("div", {
            className: "sgds-dropdown is-fullwidth is-menu-centered" + (this.state.show ? " is-active" : "")
          }, /*#__PURE__*/_react.default.createElement("div", {
            className: "sgds-dropdown-trigger"
          }, /*#__PURE__*/_react.default.createElement("button", {
            className: "sgds-button is-white is-medium is-outlined",
            "aria-haspopup": "true",
            "aria-controls": "sgds-dropdown-menu",
            onMouseEnter: function onMouseEnter() {
              return _this3.openDropdown();
            }
          }, /*#__PURE__*/_react.default.createElement("span", null, /*#__PURE__*/_react.default.createElement("b", null, this.state.selected ? this.state.selected : this.props.dropdownTitle)), /*#__PURE__*/_react.default.createElement("span", {
            className: "icon"
          }, /*#__PURE__*/_react.default.createElement("span", {
            className: "sgds-icon sgds-icon-chevron-down"
          })))), this.props.dropdownItems && this.state.show ? /*#__PURE__*/_react.default.createElement("div", {
            className: "sgds-dropdown-menu",
            role: "menu",
            onMouseLeave: function onMouseLeave() {
              return _this3.closeDropdown();
            }
          }, /*#__PURE__*/_react.default.createElement("div", {
            className: "sgds-dropdown-content"
          }, this.props.dropdownItems.map(function (item, index) {
            return /*#__PURE__*/_react.default.createElement("a", {
              href: "#",
              key: index,
              className: "sgds-dropdown-item",
              onClick: function onClick() {
                return _this3.selectDropdown(item);
              }
            }, item);
          }))) : "");
        } else {
          return "";
        }
      }
    }, {
      key: "render",
      value: function render() {
        var _this4 = this;

        return /*#__PURE__*/_react.default.createElement("section", {
          className: "sgds-hero" + (this.props.color ? "" : " has-background-primary"),
          style: {
            backgroundColor: this.props.color
          }
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-hero-body sgds-container"
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "row is-vcentered"
        }, this.props.showSearch ? /*#__PURE__*/_react.default.createElement("div", {
          className: "col is-8 is-offset-2 has-text-white has-text-centered"
        }, /*#__PURE__*/_react.default.createElement("h1", {
          className: "display"
        }, this.props.title), /*#__PURE__*/_react.default.createElement("h5", {
          className: "is-hidden-mobile margin--top--lg margin--bottom--lg"
        }, this.props.subtitle), /*#__PURE__*/_react.default.createElement("div", {
          className: "field has-addons"
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "control  is-expanded"
        }, /*#__PURE__*/_react.default.createElement("input", {
          className: "input",
          type: "text",
          placeholder: "Find a repository",
          value: this.state.searchQuery,
          onChange: this.handleSearchQuery
        })), /*#__PURE__*/_react.default.createElement("div", {
          className: "control"
        }, /*#__PURE__*/_react.default.createElement("a", {
          className: "sgds-button is-white is-outlined",
          onClick: function onClick() {
            return _this4.handleSearchButton();
          }
        }, this.props.searchButtonText ? this.props.searchButtonText : "Search")))) : /*#__PURE__*/_react.default.createElement("div", {
          className: "col is-8 is-offset-2 has-text-white has-text-centered"
        }, /*#__PURE__*/_react.default.createElement("h1", {
          className: "display"
        }, this.props.title), /*#__PURE__*/_react.default.createElement("h5", {
          className: "is-hidden-mobile margin--top--lg margin--bottom--lg"
        }, this.props.subtitle), this.props.children, this.renderDropdown()))));
      }
    }]);

    return Hero;
  }(_react.Component);

  Hero.propTypes = {
    color: _propTypes.default.string,
    title: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.object]),
    showSearch: _propTypes.default.bool
  };
  var _default = Hero;
  _exports.default = _default;
});
