// sgds-govtech-react/dist/MainNav.js
function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "styled-components", "./Brand", "prop-types"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("styled-components"), require("./Brand"), require("prop-types"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.styledComponents, global.Brand, global.propTypes);
    global.MainNav = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_exports, _react, _styledComponents, _Brand, _propTypes) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _react = _interopRequireWildcard(_react);
  _styledComponents = _interopRequireDefault(_styledComponents);
  _Brand = _interopRequireDefault(_Brand);
  _propTypes = _interopRequireDefault(_propTypes);

  var _templateObject, _templateObject2, _templateObject3;

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

  function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

  function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

  function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

  function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

  function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

  function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

  function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

  function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

  function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

  function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

  var MainNav = /*#__PURE__*/function (_Component) {
    _inherits(MainNav, _Component);

    var _super = _createSuper(MainNav);

    function MainNav(props) {
      var _this;

      _classCallCheck(this, MainNav);

      _this = _super.call(this, props);

      _defineProperty(_assertThisInitialized(_this), "selectMenuTab", function (item, index, subIndex, subMenuIndex) {
        _this.props.selectItem(item);

        if (subMenuIndex >= 0) {
          _this.setState({
            selectedTab: index,
            selectedSub: subIndex,
            selectedSubItem: subMenuIndex
          });
        } else if (subIndex >= 0) {
          _this.setState({
            selectedTab: index,
            selectedSub: subIndex,
            selectedSubItem: null
          });
        } else {
          _this.setState({
            selectedTab: index,
            selectedSub: null,
            selectedSubItem: null
          });
        }
      });

      _defineProperty(_assertThisInitialized(_this), "toggleSearchBar", function () {
        _this.setState({
          showSearch: !_this.state.showSearch
        });
      });

      _defineProperty(_assertThisInitialized(_this), "hoverOn", function (name, index, subIndex, subMenuIndex) {
        if (subMenuIndex >= 0) {
          _this.setState({
            hoverTabName: name,
            hoverTab: index,
            hoverSub: subIndex,
            hoverSunItem: subMenuIndex
          });
        } else if (subIndex >= 0) {
          _this.setState({
            hoverTabName: name,
            hoverTab: index,
            hoverSub: subIndex,
            hoverSunItem: null
          });
        } else {
          _this.setState({
            hoverTabName: name,
            hoverTab: index,
            hoverSub: null,
            hoverSunItem: null
          });
        }
      });

      _defineProperty(_assertThisInitialized(_this), "hoverOff", function () {
        _this.setState({
          hoverTabName: null,
          hoverTab: null,
          hoverSub: null,
          hoverSunItem: null
        });
      });

      _defineProperty(_assertThisInitialized(_this), "toggleMobileMenu", function () {
        _this.setState({
          expandMenu: !_this.state.expandMenu
        });
      });

      _this.state = {
        selectedTab: 0,
        selectedSub: 0,
        selectedSubItem: 0,
        hoverTabName: null,
        hoverTab: null,
        hoverSub: null,
        hoverSunItem: null,
        showSearch: false,
        expandMenu: false
      };
      _this.searchChangeHandler = props.searchChangeHandler ? props.searchChangeHandler : function () {};
      _this.searchClickHandler = props.searchClickHandler ? props.searchClickHandler : function () {};
      return _this;
    }

    _createClass(MainNav, [{
      key: "render",
      value: function render() {
        var _this2 = this;

        var NavTabWithSub = _styledComponents.default.a(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n      padding: 0.5rem 2rem 0.5rem 0.5rem;\n      align-items: center;\n      display: flex;\n      cursor: pointer;\n      color: #484848;\n      line-height: 1.5;\n      position: relative;\n      text-decoration: none;\n      font-size: inherit;\n      border-bottom: 5px solid transparent\n      &:hover {\n        font-weight: 600;\n        color: ", ";\n      }\n      &:after {\n        color: ", ";\n        content: \"\uE93C\";\n        display: block;\n        font-family: sgds-icons;\n        margin-top: -0.7rem;\n        pointer-events: none;\n        position: absolute;\n        top: 50%;\n        right: 0.5rem;\n        border: 0;\n      }\n    "])), this.props.themePrimaryColor, this.props.themePrimaryColor);

        var NavTabWithNoSub = _styledComponents.default.a(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n      border-bottom: 5px solid transparent\n      min-height: 5.25rem;\n      padding-bottom: calc(0.5rem - 1px);\n      cursor: pointer;\n      display: flex;\n      align-items: center;\n      flex-grow: 0;\n      flex-shrink: 0;\n      color: #484848;\n      line-height: 1.5;\n      padding: 0.5rem 0.75rem;\n      position: relative;\n      &:hover {\n        color: ", ";\n        font-weight: 600;\n        border-bottom: 5px solid ", ";\n      }\n    "])), this.props.themePrimaryColor, this.props.themePrimaryColor);

        var NavSub = _styledComponents.default.a(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n      font-size: 1.125rem;\n      align-items: center;\n      border: 0\n      box-sizing: border-box;\n      color: rgb(72, 72, 72);\n      cursor: pointer;\n      display: flex;\n      flex-grow: 0;\n      flex-shrink: 0;\n      height: 43px;\n      line-height: 27px;\n      margin-top: 5px;\n      padding: .5rem 1rem;\n      position: relative;\n      text-decoration: none;\n      white-space: nowrap;\n      width: 158.219px;\n      &:hover {\n        color: ", ";\n        font-weight: 600;\n      }\n    "])), this.props.themePrimaryColor);

        return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("nav", {
          className: "sgds-navbar"
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-container" + (this.props.isFluid ? " is-fluid" : "")
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-navbar-brand"
        }, /*#__PURE__*/_react.default.createElement(_Brand.default, {
          img: this.props.brand.img,
          name: this.props.brand.name,
          link: this.props.brand.link
        }), this.props.links ? /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-navbar-burger burger ".concat(this.state.expandMenu ? "is-active" : ""),
          onClick: this.toggleMobileMenu
        }, /*#__PURE__*/_react.default.createElement("span", null), /*#__PURE__*/_react.default.createElement("span", null), /*#__PURE__*/_react.default.createElement("span", null)) : null), this.props.links ? /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-navbar-menu ".concat(this.state.expandMenu ? "is-active" : "")
        }, this.props.links.map(function (link, index) {
          if (link.sublinks) {
            return /*#__PURE__*/_react.default.createElement("div", {
              className: "sgds-navbar-item has-dropdown is-hoverable",
              key: index
            }, /*#__PURE__*/_react.default.createElement(NavTabWithSub, {
              className: "is-uppercase",
              style: _this2.state.selectedTab === index ? {
                color: _this2.props.themePrimaryColor,
                fontWeight: 600,
                borderBottom: "5px solid",
                borderBottomColor: _this2.props.themePrimaryColor
              } : null
            }, link.name), /*#__PURE__*/_react.default.createElement("div", {
              className: "sgds-navbar-dropdown"
            }, link.sublinks.map(function (sublink, i) {
              return /*#__PURE__*/_react.default.createElement(NavSub, {
                href: sublink.link,
                key: i,
                style: _this2.state.selectedSub === i && _this2.state.selectedTab === index ? {
                  color: _this2.props.themePrimaryColor,
                  fontWeight: 600
                } : null,
                onClick: function onClick() {
                  return _this2.selectMenuTab(sublink.name, index, i);
                },
                onMouseEnter: function onMouseEnter() {
                  return _this2.hoverOn(sublink.name, index, i);
                },
                onMouseLeave: function onMouseLeave() {
                  return _this2.hoverOff();
                }
              }, sublink.name);
            })));
          } else if (link.subMenus) {
            return /*#__PURE__*/_react.default.createElement("div", {
              className: "sgds-navbar-item has-dropdown is-hoverable is-mega",
              key: index
            }, /*#__PURE__*/_react.default.createElement(NavTabWithSub, {
              className: "is-uppercase",
              style: _this2.state.selectedTab === index ? {
                color: _this2.props.themePrimaryColor,
                fontWeight: 600,
                borderBottom: "5px solid",
                borderBottomColor: _this2.props.themePrimaryColor
              } : null
            }, link.name), /*#__PURE__*/_react.default.createElement("div", {
              className: "sgds-navbar-dropdown"
            }, /*#__PURE__*/_react.default.createElement("div", {
              className: "sgds-container is-fluid"
            }, /*#__PURE__*/_react.default.createElement("div", {
              className: "row"
            }, link.subMenus.map(function (subMenu, i) {
              return /*#__PURE__*/_react.default.createElement("div", {
                className: "col is-3",
                key: i
              }, /*#__PURE__*/_react.default.createElement("p", {
                className: "sgds-navbar-item is-uppercase"
              }, /*#__PURE__*/_react.default.createElement("b", null, subMenu.subMenuTitle)), subMenu.subMenuItems.map(function (subMenuItem, b) {
                return /*#__PURE__*/_react.default.createElement(NavSub, {
                  key: b,
                  style: _this2.state.selectedTab === index && _this2.state.selectedSub === i && _this2.state.selectedSubItem === b ? {
                    color: _this2.props.themePrimaryColor,
                    fontWeight: 600
                  } : null,
                  onClick: function onClick() {
                    return _this2.selectMenuTab(subMenuItem.name, index, i, b);
                  },
                  onMouseEnter: function onMouseEnter() {
                    return _this2.hoverOn(subMenuItem.name, index, i, b);
                  },
                  onMouseLeave: function onMouseLeave() {
                    return _this2.hoverOff();
                  }
                }, subMenuItem.name);
              }));
            }), link.subMenuInfo ? /*#__PURE__*/_react.default.createElement("div", {
              className: "col is-6"
            }, /*#__PURE__*/_react.default.createElement("div", {
              className: "sgds-navbar-item is-wrapped"
            }, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement("b", null, link.subMenuInfo.title), /*#__PURE__*/_react.default.createElement("br", null), link.subMenuInfo.content))) : ""))));
          } else {
            return /*#__PURE__*/_react.default.createElement(NavTabWithNoSub, {
              className: "is-uppercase",
              href: link.link,
              key: index,
              style: _this2.state.selectedTab === index ? {
                color: _this2.props.themePrimaryColor,
                fontWeight: 600,
                borderBottom: "5px solid",
                borderBottomColor: _this2.props.themePrimaryColor
              } : null,
              onClick: function onClick() {
                return _this2.selectMenuTab(link.name, index);
              },
              onMouseEnter: function onMouseEnter() {
                return _this2.hoverOn(link.name, index);
              },
              onMouseLeave: function onMouseLeave() {
                return _this2.hoverOff();
              }
            }, link.img ? /*#__PURE__*/_react.default.createElement("img", {
              src: link.img,
              alt: link.name
            }) : link.name);
          }
        })) : null, this.props.displaySearch ? /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-navbar-end is-hidden-touch"
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-navbar-item"
        }, /*#__PURE__*/_react.default.createElement("button", {
          className: "sgds-button is-white is-large search-toggle",
          "data-target": "searchbar-1",
          onClick: function onClick() {
            return _this2.toggleSearchBar();
          }
        }, /*#__PURE__*/_react.default.createElement("span", {
          className: "sgds-icon sgds-icon-search"
        })))) : null)), /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-container is-hidden-touch is-marginless" + (this.state.showSearch ? null : " hide"),
          id: "searchbar-1"
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "row"
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "col"
        }, /*#__PURE__*/_react.default.createElement("hr", {
          className: "margin--bottom--sm"
        }), /*#__PURE__*/_react.default.createElement("form", null, /*#__PURE__*/_react.default.createElement("div", {
          className: "field has-addons padding--sm"
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "control has-icons-left is-expanded padding--bottom--sm"
        }, /*#__PURE__*/_react.default.createElement("input", {
          className: "input is-fullwidth is-borderless is-shadowless",
          id: "nav-4-search",
          type: "text",
          placeholder: "What are you looking for?",
          name: "nav-4-search",
          onChange: this.searchChangeHandler.bind(this)
        }), /*#__PURE__*/_react.default.createElement("span", {
          className: "icon is-left"
        }, /*#__PURE__*/_react.default.createElement("i", {
          className: "search-bar-icon sgds-icon sgds-icon-search is-size-7"
        }))), /*#__PURE__*/_react.default.createElement("div", {
          className: "control"
        }, /*#__PURE__*/_react.default.createElement("button", {
          type: "submit",
          className: "sgds-button is-primary has-text-white padding--left padding--right",
          style: {
            backgroundColor: this.props.themePrimaryColor
          },
          onClick: this.searchClickHandler.bind(this)
        }, "SEARCH"))))))));
      }
    }]);

    return MainNav;
  }(_react.Component);

  MainNav.propTypes = {
    links: _propTypes.default.arrayOf(_propTypes.default.shape({
      name: _propTypes.default.string.isRequired,
      img: _propTypes.default.string
    }))
  };
  var _default = MainNav;
  _exports.default = _default;
});
