// sgds-govtech-react/dist/Masthead.js
function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes);
    global.Masthead = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_exports, _react, _propTypes) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _react = _interopRequireWildcard(_react);
  _propTypes = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

  function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

  function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

  function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

  function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

  function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

  function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

  function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

  function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

  var Masthead = /*#__PURE__*/function (_Component) {
    _inherits(Masthead, _Component);

    var _super = _createSuper(Masthead);

    function Masthead(props) {
      var _this;

      _classCallCheck(this, Masthead);

      _this = _super.call(this, props);
      _this.hasLanguageSelector = props.hasLanguageSelector; // TODO: Test if input language is allowed

      _this.state = {
        selectedLanguage: props.defaultLanguage ? props.defaultLanguage : "English",
        isActive: false
      };
      _this.languageSelectHandler = props.languageSelectHandler;

      if (_this.languageSelectHandler) {
        _this.languageSelectHandler('English');
      } // should languages be extensible?


      _this.languages = ["English", "中文", "Bahasa Melayu", "தமிழ்"];
      return _this;
    }

    _createClass(Masthead, [{
      key: "renderLanguageList",
      value: function renderLanguageList() {
        var _this2 = this;

        return this.languages.map(function (language) {
          return language !== _this2.state.selectedLanguage ? /*#__PURE__*/_react.default.createElement("li", {
            key: language,
            alt: language,
            onClick: function onClick() {
              _this2.setState({
                selectedLanguage: language
              });

              if (_this2.languageSelectHandler) {
                _this2.languageSelectHandler(language);
              }
            }
          }, language) : "";
        });
      }
    }, {
      key: "renderLanguageSelector",
      value: function renderLanguageSelector() {
        if (this.hasLanguageSelector) {
          var style = {
            display: this.state.isActive ? "block" : "none"
          };
          return /*#__PURE__*/_react.default.createElement("div", {
            className: "language_selector--dropdown",
            style: style
          }, /*#__PURE__*/_react.default.createElement("ul", null, this.renderLanguageList()));
        }
      }
    }, {
      key: "toggleLanguageSelector",
      value: function toggleLanguageSelector() {
        this.setState({
          isActive: !this.state.isActive
        });
      }
    }, {
      key: "renderSelectedLanguage",
      value: function renderSelectedLanguage() {
        if (this.hasLanguageSelector) {
          return /*#__PURE__*/_react.default.createElement("div", {
            onClick: this.toggleLanguageSelector.bind(this),
            className: "col is-flex has-text-right language_selector"
          }, this.state.selectedLanguage);
        }
      }
    }, {
      key: "render",
      value: function render() {
        return /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-masthead"
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "sgds-container"
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "row"
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: "col is-9"
        }, /*#__PURE__*/_react.default.createElement("a", {
          href: "https://www.gov.sg"
        }, /*#__PURE__*/_react.default.createElement("span", {
          className: "sgds-icon sgds-icon-sg-crest"
        }), /*#__PURE__*/_react.default.createElement("span", {
          className: "is-text"
        }, "A Singapore Government Agency Website"))))));
      }
    }]);

    return Masthead;
  }(_react.Component);

  Masthead.propTypes = {
    hasLanguageSelector: _propTypes.default.bool
  };
  Masthead.defaultProps = {
    onClick: function onClick() {}
  };
  var _default = Masthead;
  _exports.default = _default;
});
