// sgds-govtech-react/dist/Notification.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes);
    global.Notification = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_exports, _react, _propTypes) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _react = _interopRequireDefault(_react);
  _propTypes = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

  function Notification(props) {
    var getSelfClassName = function getSelfClassName() {
      var _props$className;

      return (_props$className = props.className) !== null && _props$className !== void 0 ? _props$className : false;
    };

    var getClassName = function getClassName() {
      var className = "";

      if (props.isToast) {
        className = "sgds-notification is-toast";
      } else {
        className = "sgds-notification";
      }

      if (props.colour) {
        className = "".concat(className, " is-").concat(props.colour);
      }

      if (getSelfClassName()) {
        return "".concat(className, " ").concat(getSelfClassName());
      } else {
        return className;
      }
    };

    var getIcon = function getIcon() {
      return "sgds-icon ".concat(props.icon, " is-size-4");
    };

    return /*#__PURE__*/_react.default.createElement("div", {
      className: getClassName()
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "sgds-notification-detail"
    }, props.icon ? /*#__PURE__*/_react.default.createElement("span", {
      className: getIcon()
    }) : null, /*#__PURE__*/_react.default.createElement("div", {
      className: "sgds-notification-content"
    }, /*#__PURE__*/_react.default.createElement("p", {
      className: "has-text-weight-bold"
    }, props.title), /*#__PURE__*/_react.default.createElement("p", null, props.content)), props.closable ? /*#__PURE__*/_react.default.createElement("span", {
      className: "delete",
      onClick: function onClick() {
        props.onClose();
      }
    }) : null));
  }

  Notification.propTypes = {
    isToast: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.string]),
    title: _propTypes.default.string,
    content: _propTypes.default.string,
    icon: _propTypes.default.string,
    closable: _propTypes.default.bool,
    onClose: _propTypes.default.func,
    colour: _propTypes.default.string,
    style: _propTypes.default.object,
    className: _propTypes.default.string
  };
  var _default = Notification;
  _exports.default = _default;
});
