// sgds-govtech-react/dist/Row.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes);
    global.Row = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_exports, _react, _propTypes) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _react = _interopRequireDefault(_react);
  _propTypes = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

  function Row(props) {
    var getSelfClassName = function getSelfClassName() {
      var _props$className;

      return (_props$className = props.className) !== null && _props$className !== void 0 ? _props$className : false;
    };

    var getClassName = function getClassName() {
      var className = "";

      if (props.isDesktop) {
        className = "row is-desktop";
      } else if (props.isMobile) {
        className = "row is-mobile";
      } else {
        className = "row";
      }

      if (props.isMultiline) {
        className = className + " is-multiline";
      }

      if (getSelfClassName()) {
        return "".concat(className, " ").concat(getSelfClassName());
      } else {
        return className;
      }
    };

    return /*#__PURE__*/_react.default.createElement("div", {
      className: getClassName(),
      style: props.style
    }, props.children);
  }

  Row.propTypes = {
    isMultiline: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.string]),
    isDesktop: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.string]),
    isMobile: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.string]),
    style: _propTypes.default.object,
    className: _propTypes.default.string
  };
  var _default = Row;
  _exports.default = _default;
});
