// sgds-govtech-react/dist/Section.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes);
    global.Section = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_exports, _react, _propTypes) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _react = _interopRequireDefault(_react);
  _propTypes = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

  function Section(props) {
    var getSelfClassName = function getSelfClassName() {
      var _props$className;

      return (_props$className = props.className) !== null && _props$className !== void 0 ? _props$className : false;
    };

    var getClassName = function getClassName() {
      var className = "";

      if (props.isSmall) {
        className = "sgds-section is-small";
      } else if (props.isMedium) {
        className = "sgds-section is-medium";
      } else if (props.isLarge) {
        className = "sgds-section is-large";
      } else {
        className = "sgds-section";
      }

      if (getSelfClassName()) {
        return "".concat(className, " ").concat(getSelfClassName());
      } else {
        return className;
      }
    };

    return /*#__PURE__*/_react.default.createElement("section", {
      className: getClassName(),
      style: props.style
    }, props.children);
  }

  Section.propTypes = {
    isSmall: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.string]),
    isMedium: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.string]),
    isLarge: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.string]),
    style: _propTypes.default.object,
    className: _propTypes.default.string
  };
  var _default = Section;
  _exports.default = _default;
});
