// sgds-govtech-react/dist/standard/Masthead.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes);
    global.Masthead = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_exports, _react, _propTypes) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.Masthead = Masthead;
  _react = _interopRequireDefault(_react);
  _propTypes = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

  function MastheadContent(_ref) {
    var children = _ref.children;
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
      className: "sgds-icon sgds-icon-sg-crest"
    }), /*#__PURE__*/_react.default.createElement("span", {
      className: "is-text"
    }, children));
  }

  function Masthead(_ref2) {
    var _ref2$isFluid = _ref2.isFluid,
        isFluid = _ref2$isFluid === void 0 ? false : _ref2$isFluid,
        _ref2$noLink = _ref2.noLink,
        noLink = _ref2$noLink === void 0 ? false : _ref2$noLink,
        children = _ref2.children;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "sgds-masthead"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "sgds-container ".concat(isFluid ? "is-fluid" : "")
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "row"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "col"
    }, noLink ? /*#__PURE__*/_react.default.createElement(MastheadContent, null, children || "A Singapore Government Agency Website") : /*#__PURE__*/_react.default.createElement("a", {
      href: "https://www.gov.sg"
    }, /*#__PURE__*/_react.default.createElement(MastheadContent, null, children || "A Singapore Government Agency Website"))))));
  }

  Masthead.propTypes = {
    isFluid: _propTypes.default.bool,
    noLink: _propTypes.default.bool
  };
});
